### PR TITLE
[MXNET-33] SSD example not working with mkl-dnn

### PR DESCRIPTION
**Pull request by ashokei**
## Description ##
SSD VGG16-reduced example is not working if pooling convention is "full" with Max pool. This PR resolves the this issue by defaulting to CPU implementation when pooling convention is "full". 

### Essentials ###
- [ ✅ ] Passed code style checking (`make lint`)
- [ ✅ ] Changes are complete (i.e. I finished coding on this PR)

### Changes ###
- disable mkl-dnn if pooling convention is "full"

## Comments ##
- This PR is to resolve the crash seen. A new PR may be needed to implement "full" pooling convention with mkl-dnn.

### CODE REVIEW

**Comment by ashokei** [comment]
@zheng-da @TaoLv please review. I notice we do not have proper "full" pooling convention support, this PR disables mkl-dnn pooling for "full" convention; 

We will probably need a new PR to implement the "full" convention feature with mkl-dnn for both Max and Avg Pooling cases.

**Comment by TaoLv** [comment]
There is a test case for max pooling with `full` convention: https://github.com/apache/incubator-mxnet/blob/master/tests/python/gpu/test_operator_gpu.py#L841 
But it isn't observed that this case failed before. If the test case doesn't cover the corner case you mentioned, maybe we need add new test case for it.

File: src/operator/nn/mkldnn/mkldnn_pooling-inl.h [hunk 1]
```diff
@@ -92,6 +92,8 @@ inline bool SupportMKLDNNPooling(const PoolingParam &param,
 
   if (param.pooling_convention == pool_enum::kValid)
     return true;
+  else
+    return false;
```
Comment (piiswrong) [inline src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: this makes the later code unreachable. Are you sure this is what you want?

Comment (ashokei) [reply src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: @piiswrong the later code is being disabled for now until we support all "full" pooling convention cases.  As @TaoLv mentioned, i will add a unit-test for this failure. 
We may refactor this code, so marking this PR WIP for now.

Comment (zheng-da) [reply src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: maybe use "#if 0" to comment the checks below?

**Comment by ashokei** [comment]
@marcoabreu @TaoLv @zheng-da Unit-tests currently do not seem to test backward/training mode for pooling very well. Below is the test for reproducing this crash.

`
cd example/image-classification/symbols/
ln -s ../../ssd/symbol/vgg16_reduced.py vgg16_reduced.py
cd ../../..
python example/image-classification/train_cifar10.py --network vgg16_reduced
`

This PR will fix the above crash.

**Comment by marcoabreu** [comment]
Feel free to develop and add an appropriate unit test yourself. Please don't hesitate to reach out to us in case you need assistance.

**Comment by ashokei** [comment]
@marcoabreu this issue only happens inside a particular model (SSD VGG16-reduced), i clarified in title/description. I'm not sure a special unittest just for this specific model is appropriate. I am not able to reproduce this issue on separately. please let me know if any other suggestions/feedback, thanks!.

**Comment by marcoabreu** [comment]
That's not an issue, we're already testing a variety of models and the bigger diversity, the better. We can re-visit this decision later on and reduce the test size if required. For now, it would be good to get a test since we would like to re-enable pooling for MKL-DNN at a later point in time - this test would give us a tool to verify that it's actually working again.

File: tests/python/cpu/test_mkldnn.py [hunk 2]
```diff
@@ -41,18 +42,49 @@ def test_mkldnn_install():
...
+            "variable settings"
+
+
+def test_mkldnn_model():
+    """
+    This test will run a sample model for couple of iterations.
+    """
+
+    import mxnet as mx
+    model = os.path.join(os.path.dirname(os.path.realpath(__file__)), "test_model.json")
```
Comment (marcoabreu) [inline tests/python/cpu/test_mkldnn.py]: could you move this file to a subdirectory like tests/python/cpu/data/test_mkldnn_ test_mkldnn_model_model1.json? (format is %FILENAME%_%TESTNAME%_model%NUMBER%.json)?

File: tests/python/cpu/test_mkldnn.py [hunk 3]
```diff
@@ -41,18 +42,49 @@ def test_mkldnn_install():
...
         logging.info("MXNet is built/installed correctly with MKL-DNN")
     else:
         assert 0, "MXNet is built/installed incorrectly with MKL-DNN, please " \
-               "double check your build/install steps or environment " \
-               "variable settings"
+            "double check your build/install steps or environment " \
+            "variable settings"
+
+
+def test_mkldnn_model():
```
Comment (marcoabreu) [inline tests/python/cpu/test_mkldnn.py]: Please explain that this test has no asserts because it is only supposed to run successfully without failure.

Comment (ashokei) [reply tests/python/cpu/test_mkldnn.py]: @marcoabreu i added assert for sanity check, and renamed file as you mentioned. any other suggestions , please let me know.

File: src/operator/nn/mkldnn/mkldnn_pooling-inl.h [hunk 4]
```diff
@@ -92,12 +92,17 @@ inline bool SupportMKLDNNPooling(const PoolingParam &param,
...
 
+// need to support pooling convention full
```
Comment (marcoabreu) [inline src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: Please add a link to the issue as part of the comment

File: tests/python/cpu/test_mkldnn.py (hunk 3 above)
Comment (marcoabreu) [reply tests/python/cpu/test_mkldnn.py]: Thanks!

File: src/operator/nn/mkldnn/mkldnn_pooling-inl.h (hunk 4 above)
Comment (ashokei) [reply src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: you mean , link to this PR ? or JIRA ?

Comment (marcoabreu) [reply src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: To jira

**Comment by pengzhao-intel** [comment]
#10092 
@marcoabreu could you help review again and merge?

**Comment by zheng-da** [comment]
@ashokei @pengzhao-intel do you know when the Intel MKLDNN team will fix the bug in MKLDNN?

File: tests/python/cpu/test_mkldnn.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (marcoabreu) [inline tests/python/cpu/test_mkldnn.py]: tests in ```tests/python/cpu``` are currently not run in CI. Please add it to the MKLDNN jobs.

**Comment by ashokei** [comment]
@zheng-da This is not a bug in MKL-DNN, it is an issue with current MKL-DNN mxnet integration. This PR will use CPU fallback.

File: tests/python/cpu/test_mkldnn.py (hunk 5 above)
Comment (ashokei) [reply tests/python/cpu/test_mkldnn.py]: @marcoabreu This PR only adds to existing test, no new test file is created. Can you be specific which CI jobs you are referring to ? If there is a gap in CI, a new JIRA item can be created to address it.

**Comment by zheng-da** [comment]
what i mean is that mkldnn doesn't support "full" pooling convention. will mkldnn eventually support it?

**Comment by zheng-da** [comment]
oh, i see. do you mind implementing it? fallback is fine, but it'll be better if it can invoke the one in MKLDNN.

**Comment by ashokei** [comment]
yes, we plan to add new features for pooling implementation with mkl-dnn in future PR.
this PR just fixes the crash in SSD, and this specific case (maxpool with fullconvention) is only used in 1 layer, and not common.

File: tests/python/cpu/test_mkldnn.py (hunk 5 above)
Comment (marcoabreu) [reply tests/python/cpu/test_mkldnn.py]: According to https://github.com/apache/incubator-mxnet/blob/master/Jenkinsfile#L439, which in the end points to https://github.com/apache/incubator-mxnet/blob/master/ci/docker/runtime_functions.sh#L387, we're not running the tests located in tests/python/cpu. Please rename tests/python/cpu to tests/python/mkl and add another function in the same style as https://github.com/apache/incubator-mxnet/blob/master/Jenkinsfile#L458.

**Comment by ashokei** [comment]
@marcoabreu does this look ok, thanks.

**Comment by marcoabreu** [comment]
I think something just broke, I'm seeing 218 changed files and a lot of unrelated changes. Could you please try to rebase and ping me afterwards? Thank you!

**Comment by ashokei** [comment]
@marcoabreu fixed now

**Comment by anirudh2290** [comment]
@ashokei can you  please check if the CI failure is related to your change and retrigger it if not ?

**Comment by ashokei** [comment]
@anirudh2290 done...thanks.

**Comment by zheng-da** [comment]
@piiswrong this PR is to fix a bug in the MKLDNN integration. MKLDNN pooling can't handle full pooling convention correctly.
do you have any comments for this PR? if not, can you merge it?

**Comment by ashokei** [comment]
@piiswrong can you please merge this, this PR also has mkldnn jenkins/unittest as requested by @marcoabreu

**Pull request merged**